cms/src/symmetric_key.rs

```diff
@@ -0,0 +1,237 @@
+///        sKeyAttrs  SEQUENCE SIZE (1..MAX) OF Attribute
+///                                       {{ SKeyAttributes }}  OPTIONAL,
+///        sKey       OCTET STRING OPTIONAL }
+///        ( WITH COMPONENTS { ..., sKeyAttrs PRESENT } |
+///          WITH COMPONENTS { ..., sKey PRESENT } )
+/// ```
+///
+/// [RFC 6031 Section 2.0]: https://datatracker.ietf.org/doc/html/rfc6031#section-2
+#[derive(Sequence, PartialEq, Eq)]
+#[allow(missing_docs)]
+pub struct OneSymmetricKey {
+    pub s_key_attrs: Vec<Attribute>,
+    #[asn1(optional = "true")]
+    pub s_key: Option<OctetString>,
+}
+
+/// The `KeyPkgVersion` type is defined in [RFC 6031 Section 2.0].
+///
+/// ```text
+///     KeyPkgVersion ::= INTEGER  { v1(1) } ( v1, ... )
+/// ```
+/// [RFC 6031 Section 2.0]: https://datatracker.ietf.org/doc/html/rfc6031#section-2
+#[derive(Default, Clone, Debug, Copy, PartialEq, Eq, PartialOrd, Ord, Enumerated)]
+#[asn1(type = "INTEGER")]
+#[repr(u8)]
+#[allow(missing_docs)]
+pub enum KeyPkgVersion {
+    #[default]
+    V1 = 1,
+}
+
+// todo: determine if all the types below should live in another file such as attr.rs
+
+/// The `FriendlyName` type is defined in [RFC 6031 Section 3.2.6].
+///
+/// ```text
+///    FriendlyName ::= SEQUENCE {
+///      friendlyName        UTF8String,
+///      friendlyNameLangTag UTF8String OPTIONAL }
+/// ```
+///
+/// [RFC 6031 Section 3.2.6]: https://datatracker.ietf.org/doc/html/rfc6031#section-3.2.6
+#[derive(Sequence, PartialEq, Eq)]
+#[allow(missing_docs)]
+pub struct FriendlyName {
+    pub friendly_name: String,
+    #[asn1(optional = "true")]
+    pub friendly_name_lang_tag: Option<String>,
+}
+
+/// The `PSKCAlgorithmParameters` type is defined in [RFC 6031 Section 3.2.7].
+///
+///  ```text
+///    PSKCAlgorithmParameters ::= CHOICE {
+///      suite                UTF8String,
+///      challengeFormat  [0] ChallengeFormat,
+///      responseFormat   [1] ResponseFormat,
+///      ... }
+/// ```
+///
+/// [RFC 6031 Section 3.2.7]: https://datatracker.ietf.org/doc/html/rfc6031#section-3.2.7
+#[derive(Sequence, PartialEq, Eq)]
+#[allow(missing_docs)]
+pub struct PSKCAlgorithmParameters<'range> {
+    pub suite: String,
+    pub challenge_format: ChallengeFormat<'range>,
+    pub response_format: ResponseFormat,
+}
+
+/// The `ChallengeFormat` type is defined in [RFC 6031 Section 3.2.7].
+///
+/// ```text
+///    ChallengeFormat ::= SEQUENCE {
+///      encoding    Encoding,
+///      checkDigit  BOOLEAN DEFAULT FALSE,
+///      min         INTEGER (0..MAX),
+///      max         INTEGER (0..MAX),
+///      ... }
+/// ```
+///
+/// [RFC 6031 Section 3.2.7]: https://datatracker.ietf.org/doc/html/rfc6031#section-3.2.7
+#[derive(Sequence, PartialEq, Eq)]
+#[allow(missing_docs)]
+pub struct ChallengeFormat<'range> {
+    pub encoding: Encoding,
+    #[asn1(default = "Default::default")]
+    pub check_digit: bool,
+    pub min: der::asn1::IntRef<'range>,
+    pub max: der::asn1::IntRef<'range>,
```

Comment:
You probably can use the owned version of int:
```suggestion
pub struct ChallengeFormat {
    pub encoding: Encoding,
    #[asn1(default = "Default::default")]
    pub check_digit: bool,
    pub min: der::asn1::Int,
    pub max: der::asn1::Int,
```